### ISSUE TITLE
ICE on invalid code

### ISSUE THREAD

**Issue by ctennis** [rust-lang/rust#20585]
I know this is invalid code, but I was experimenting with syntax, when generating this ice:

```rust
        // line is a String
        let ss: Vec<String> = line.split(' ').map(|s| s.to_string()).collect();  // ok
        let ss2 = line.split("asdf").collect()[1];  // throws ICE
        // specifically, the addition of the [1] on the end of the above line causes the ICE
```

```
rustc 0.13.0-nightly (ad9e75938 2015-01-05 00:26:28 +0000)
binary: rustc
commit-hash: ad9e759382ad7daed26f86732f41f5f83cd673e2
commit-date: 2015-01-05 00:26:28 +0000
host: x86_64-apple-darwin
release: 0.13.0-nightly
```

```
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libsyntax/diagnostic.rs:182

stack backtrace:
   1:        0x110ebdb55 - sys::backtrace::write::hf19ca6a05eb5ebb2eTs
   2:        0x110ee34bf - failure::on_fail::h7c18013e637fb30dGVy
   3:        0x110e489ba - rt::unwind::begin_unwind_inner::h529f368fda5064a6GDy
   4:        0x10eb52cd7 - rt::unwind::begin_unwind::h13948472927873582381
   5:        0x10eb53648 - diagnostic::Handler::bug::hd71c943ddf83830fDLF
   6:        0x10e16cdd8 - session::Session::bug::hd47af5d46084797cFoq
   7:        0x10dadf7f2 - check::FnCtxt<'a, 'tcx>::node_ty::h2d96c1efc4f011a5EJl
   8:        0x10db60e88 - middle::mem_categorization::MemCategorizationContext<'t, TYPER>::cat_expr_unadjusted::h16076163679649109229
   9:        0x10db61b23 - middle::mem_categorization::MemCategorizationContext<'t, TYPER>::cat_expr::h4303509986251718759
  10:        0x10db624d6 - middle::expr_use_visitor::ExprUseVisitor<'d, 't, 'tcx, TYPER>::consume_expr::h14848432851457813560
  11:        0x10db5f561 - middle::expr_use_visitor::ExprUseVisitor<'d, 't, 'tcx, TYPER>::walk_expr::h2141893513736095856
  12:        0x10db5e167 - middle::expr_use_visitor::ExprUseVisitor<'d, 't, 'tcx, TYPER>::walk_block::h7274518944735195104
  13:        0x10db5ec3d - middle::expr_use_visitor::ExprUseVisitor<'d, 't, 'tcx, TYPER>::walk_expr::h2141893513736095856
  14:        0x10db6259c - middle::expr_use_visitor::ExprUseVisitor<'d, 't, 'tcx, TYPER>::consume_expr::h14848432851457813560
  15:        0x10db5e24a - middle::expr_use_visitor::ExprUseVisitor<'d, 't, 'tcx, TYPER>::walk_block::h7274518944735195104
  16:        0x10db55d61 - check::upvar::AdjustBorrowKind<'a, 'tcx>::analyze_fn::h5bef29ec6deeacaeUUh
  17:        0x10dbc9993 - check::check_bare_fn::heedebd2b9e51b0dePCj
  18:        0x10dbc068d - check::check_item::h5f6ef4508fad905dBVj
  19:        0x10dd4bf30 - check_crate::unboxed_closure.40253
  20:        0x10dd47793 - check_crate::hd84c1363050258aeiux
  21:        0x10d5a28c2 - driver::phase_3_run_analysis_passes::h8359258d16ec196cSva
  22:        0x10d58f24c - driver::compile_input::hf91a1679406c88e7vba
  23:        0x10d656863 - thunk::F.Invoke<A, R>::invoke::h3846886491688616995
  24:        0x10d6539c0 - rt::unwind::try::try_fn::h18042905528533303549
  25:        0x110f481c9 - rust_try_inner
  26:        0x110f481b6 - rust_try
  27:        0x10d654106 - thunk::F.Invoke<A, R>::invoke::h10752720603711681108
  28:        0x110ecefa4 - sys::thread::thread_start::h864a1c4bf92ed975gGv
  29:     0x7fff95c5c2fc - _pthread_body
  30:     0x7fff95c5c279 - _pthread_body
```

**Comment by drawoc** [comment]
This looks similar to #20614 to me.

**Comment by ctennis** [comment]
Agreed, this is fixed by the same merge that fixed #20614
